### PR TITLE
Subject identifier (sub) replaced email for user id (backend internal)

### PR DESCRIPTION
**Pull request by aiAdrian**
# Description

# Issues

# Checklist

* [x] This PR contains a description of the changes I'm making
* [x] I've read the [Contribution Guidelines](https://github.com/SchweizerischeBundesbahnen/netzgrafik-editor-backend/blob/main/CONTRIBUTING.md)
* [x] I've added tests for changes or features I've introduced
* [x] CI is currently green and this is ready for review

### CODE REVIEW

**Review by chenkins — CHANGES_REQUESTED** [review]
@aiAdrian as discussed, the existing projects will become unaccessible for existing users through this change.  The `AuthenticationService:.getCurrentUserId()` is used by the `ProjectService` etc. So two comments:
* It would be more proper to continue using the internal sub and to only display the corresponding e-mail in the sharing dialogue. Usually, frontend frameworks support this distinction. 
* In addition, it it is possible that the same e-mail is used for multiple social login providers - this could lead to some edge cases you must decide to either test or ignore.

**Comment by aiAdrian** [comment]
> @aiAdrian as discussed, the existing projects will become unaccessible for existing users through this change. The `AuthenticationService:.getCurrentUserId()` is used by the `ProjectService` etc. So two comments:
> 
> * It would be more proper to continue using the internal sub and to only display the corresponding e-mail in the sharing dialogue. Usually, frontend frameworks support this distinction.
> * In addition, it it is possible that the same e-mail is used for multiple social login providers - this could lead to some edge cases you must decide to either test or ignore.

@chenkins as i could understand, if you want to map the sub identifier to the corresponding email address, it would need an additional data field to store this mapping (database) and the filtering must be based on this field. 

I see the point that the same e-mail could be used. This would lead to break the concept one login - one "database" view. This is very good point. But on the other hand - if we like to share data through email adresses this would as well lead to an edge case when the diffrent social logins have used same e-mail.

File: src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
 public class ProjectController {
     private final ProjectService projectService;
 
-    private static Pattern USER_ID_PATTERN = Pattern.compile("^(u|ue|e)\\d+$");
+    private static Pattern USER_ID_PATTERN = Pattern.compile("^([a-z0-9_\\.-]+)@([\\da-z\\.-]+)\\.([a-z\\.]{2,6})$");
```
[the remark this answers is not in the document]
Comment (aiAdrian) [reply src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java]: Done

**Comment by aiAdrian** [comment]
@chenkins  : Thank for the review. 

I changed 
- Renamed / refactored
- Added tests
- Added small documentation to the reg exp.

File: src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java [hunk 2]
```diff
@@ -35,7 +35,8 @@
 public class ProjectController {
     private final ProjectService projectService;
 
-    private static Pattern USER_ID_PATTERN = Pattern.compile("^(u|ue|e)\\d+$");
+    // email adresse validator: regex to match emails using the expression
```
Comment (chenkins) [inline src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java:38]: typo
```suggestion
    // email adress validator: regex to match emails using the expression
```

**Review by chenkins — APPROVED** [review]

File: src/test/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        assertFalse(isUserIdAsEmailPatternValid("u123456"),"u123456 should be invalid");
+        assertFalse(isUserIdAsEmailPatternValid("name.vorname.vorname2#mail.domain.ch"),"name.vorname.vorname2#mail.domain.ch should be invalid");
+        assertFalse(isUserIdAsEmailPatternValid(""),"'' should be invalid");
+        assertFalse(isUserIdAsEmailPatternValid("Adrian@ai.org"),"adrian@ai.org should be valid");
+        assertFalse(isUserIdAsEmailPatternValid("aDrian@ai.org"),"adrian@ai.org should be valid");
+        assertFalse(isUserIdAsEmailPatternValid("ADRIAN@AI.ORG"),"adrian@ai.org should be valid");
+        assertFalse(isUserIdAsEmailPatternValid("adrian@AI.ORG"),"adrian@ai.org should be valid");
+        assertFalse(isUserIdAsEmailPatternValid("adrian@ai.ORG"),"adrian@ai.org should be valid");
+        assertFalse(isUserIdAsEmailPatternValid("adrian@AI.org"),"adrian@ai.org should be valid");
+        assertFalse(isUserIdAsEmailPatternValid("adrian@ai.orG"),"adrian@ai.org should be valid");
```
Comment (chenkins) [inline src/test/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectControllerTest.java:42]: message incorrect, should be `invalid`. Also, cases incorrect.

Comment (aiAdrian) [reply src/test/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectControllerTest.java]: Thank you - fixed

File: src/test/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        assertTrue(isUserIdAsEmailPatternValid("name.vorname.vorname2@mail.domain.ch"),"name.vorname.vorname2@mail.domain.ch should be valid");
+        assertTrue(isUserIdAsEmailPatternValid("fun@data.cloud"),"fun@data.cloud should be valid");
+        assertTrue(isUserIdAsEmailPatternValid("adrian@ai.org"),"adrian@ai.org should be valid");
+        assertTrue(isUserIdAsEmailPatternValid("1234@x.org"),"1234@x.org should be valid");
+        assertTrue(isUserIdAsEmailPatternValid("x@1234.org"),"x@1234.org should be valid");
+        assertTrue(isUserIdAsEmailPatternValid("1234@1234.org"),"1234@1234.org should be valid");
+        assertTrue(isUserIdAsEmailPatternValid("123a4@1234.org"),"123a4@1234.org should be valid");
+        assertTrue(isUserIdAsEmailPatternValid("123a4@1234abc123zsd.a123sb.org"),"123a4@1234.org should be valid");
+
+        assertFalse(isUserIdAsEmailPatternValid("Franz@nix.com"),"Franz@nix.com should be invalid");
```
Comment (emersion) [inline src/test/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectControllerTest.java:31]: This is a valid e-mail address.

File: src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java [hunk 5]
```diff
@@ -35,7 +35,8 @@
 public class ProjectController {
     private final ProjectService projectService;
 
-    private static Pattern USER_ID_PATTERN = Pattern.compile("^(u|ue|e)\\d+$");
+    // email adress validator: regex to match emails using the expression
+    protected static Pattern USER_ID_AS_EMAIL_PATTERN = Pattern.compile("^([a-z0-9_\\.-]+)@([a-z0-9\\.-]+).([a-z\\.]{2,6})$");
```
[the remark this answers is not in the document]
Comment (aiAdrian) [reply src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java:39]: excellent. Thank you

[the remark this answers is not in the document]
Comment (aiAdrian) [reply src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java:39]: Thanks for the comment: changed the pattern. 

As well added the enforcing storing of email addresses (user IDs) in lowercase and comparing them with lowercase strings, this ensures consistency and eliminate any potential discrepancies caused by case sensitivity.

[the remark this answers is not in the document]
Comment (aiAdrian) [reply src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java]: id.match(/^([a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*)|(u|ue|e)\d+$/);

File: src/test/java/ch/sbb/pfi/netzgrafikeditor/integrationtest/CreateProjectIT.java [hunk 6]
```diff
@@ -95,6 +95,9 @@ public void createAndFetchProjects__expectCreatedProject() throws Exception {
                 .andExpect(status().isOk())
                 .andExpect(jsonResponse(expectedResponse.withIsWritable(false)));
 
+
+        System.out.println('ADRIAN');
+        System.out.println(location);
```
[the remark this answers is not in the document]
Comment (aiAdrian) [reply src/test/java/ch/sbb/pfi/netzgrafikeditor/integrationtest/CreateProjectIT.java]: done. thanks . artefakt

File: src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (aiAdrian) [reply src/main/java/ch/sbb/pfi/netzgrafikeditor/api/project/ProjectController.java]: flatten? i changed the pattern.
